src/Agent.Worker/TestResults/ResultsCommandExtension.cs

```diff
@@ -181,43 +200,85 @@ private void ProcessPublishTestResultsCommand(IExecutionContext context, Diction
                     }
 
                     minStartDate = DateTime.Equals(minStartDate, DateTime.MaxValue) ? presentTime : minStartDate;
-                    maxCompleteDate = dateFormatError || DateTime.Equals(maxCompleteDate, DateTime.MinValue) ? minStartDate.Add(totalTestCaseDuration) : maxCompleteDate;
-
-                    //creat test run
-                    TestRunData testRunData = new TestRunData(
-                        name: runName,
-                        startedDate: minStartDate.ToString("o"),
-                        completedDate: maxCompleteDate.ToString("o"),
-                        state: "InProgress",
-                        isAutomated: true,
-                        buildId: runContext != null ? runContext.BuildId : 0,
-                        buildFlavor: runContext != null ? runContext.Configuration : string.Empty,
-                        buildPlatform: runContext != null ? runContext.Platform : string.Empty,
-                        releaseUri: runContext != null ? runContext.ReleaseUri : null,
-                        releaseEnvironmentUri: runContext != null ? runContext.ReleaseEnvironmentUri : null
-                    );
-
-                    testRunData.Attachments = runAttachments.ToArray();
-
-                    TestRun testRun = await publisher.StartTestRunAsync(testRunData, _executionContext.CancellationToken);
-                    await publisher.AddResultsAsync(testRun, runResults.ToArray(), _executionContext.CancellationToken);
-                    await publisher.EndTestRunAsync(testRunData, testRun.Id, true, _executionContext.CancellationToken);
+                    maxCompleteDate = dateFormatError || DateTime.Equals(maxCompleteDate, DateTime.MinValue)
+                        ? minStartDate.Add(totalTestCaseDuration)
+                        : maxCompleteDate;
+
+
+                    var testRunDataEnd =
+                        new TestRunData(startedDate: minStartDate.ToString("o"),
+                            completedDate: maxCompleteDate.ToString("o")) {Attachments = runAttachments.ToArray()};
+                    TestRun testRun = await testRunInitializationAsyncLazy.Value;
+                    await publisher.EndTestRunAsync(testRunDataEnd, testRun.Id, true, _executionContext.CancellationToken);
```

Comment:
`cancellationToken` instead of `_executionContext.CancellationToken`, make sure you use the same cancellation token in all the place.